odoo/addons/base/models/ir_default.py

```diff
@@ -188,3 +207,103 @@ def discard_values(self, model_name, field_name, values):
         json_vals = [json.dumps(value, ensure_ascii=False) for value in values]
         domain = [('field_id', '=', field.id), ('json_value', 'in', json_vals)]
         return self.search(domain).unlink()
+
+    def _evaluate_leaf_with_fallback(self, model_name, leaf):
```

Comment:
I think that removing the **fallback** value from company_dependent field will simplify a lot of the code here. If you need some fallback, you may just use computed fields.
```
some_field = fields.Char(computed='as `value_for_company or value_fallback`')
value_for_company = fields.Char(company_dependent=True)
# fallback can be a computed field,
# or could be a function on the model to be called or could even read ir_default
value_fallback = fields.Char()
```
You would avoid having to add fallbacks to expression and ir_default would not need to invalidate everything anymore.

Reply:
But the logic company dependent field wiht fallback is need in many places, and they may need to be searched. When you implement the `_search` or `_field_to_sql` you still need such kind of method to tell if fallback will be filtered out by the domain

Reply:
Yes, but you can implement the fallback logic case-by-case making overall implementation simpler.

You will also be able to distinguish cases when a value in X because it is set to X on the record for a given company or because it is a fallback value. Even for developers and for debugging, it will be clearer.
```
# suppose you have  a default and 2 records
default = X
rec1 = {'cdf': {1: X, 2: Y}}
rec2 = {'cdf': null}
# if you change the default to Z, then rec2 changes, but rec1 does not change
```

Reply:
After today's discussion we can implement the fallback as it is today.